src/libraries/Microsoft.Extensions.Options.ConfigurationExtensions/src/OptionsBuilderConfigurationExtensions.cs

```diff
@@ -40,5 +40,43 @@ public static class OptionsBuilderConfigurationExtensions
             optionsBuilder.Services.Configure<TOptions>(optionsBuilder.Name, config, configureBinder);
             return optionsBuilder;
         }
+
+        /// <summary>
+        /// Registers the dependency injection container to bind <typeparamref name="TOptions"/> against
+        /// the <see cref="IConfiguration"/> obtained from the DI service provider.
+        /// </summary>
+        /// <typeparam name="TOptions">The options type to be configured.</typeparam>
+        /// <param name="optionsBuilder">The options builder to add the services to.</param>
+        /// <param name="configSectionPath">
+        /// The name of the configuration section to bind from. If omitted, or
+        /// <see langword="null"/> the name of the <paramref name="optionsBuilder"/>
+        /// is used.
+        /// </param>
+        /// <param name="configureBinder">Optional. Used to configure the <see cref="BinderOptions"/>.</param>
+        /// <returns>The <see cref="OptionsBuilder{TOptions}"/> so that additional calls can be chained.</returns>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="optionsBuilder"/> is <see langword="null"/>.
+        /// </exception>
+        /// <seealso cref="Bind{TOptions}(OptionsBuilder{TOptions}, IConfiguration, Action{BinderOptions})"/>
+        public static OptionsBuilder<TOptions> BindConfiguration<TOptions>(
+            this OptionsBuilder<TOptions> optionsBuilder,
+            string configSectionPath = null,
+            Action<BinderOptions> configureBinder = null)
+            where TOptions : class
+        {
+            _ = optionsBuilder ?? throw new ArgumentNullException(nameof(optionsBuilder));
+
+            optionsBuilder.Configure<IConfiguration>((opts, config) =>
+            {
+                IConfiguration section = null;
+                if (string.IsNullOrEmpty(configSectionPath))
+                    section = config?.GetSection(configSectionPath);
+                else if (!string.IsNullOrEmpty(optionsBuilder.Name))
```

Comment:
I don't think this makes sense. I was thinking maybe `nameof(TOptions)` as a convention but it would mean that the root would never be chosen. The idea being that it's rarely (if ever) useful to bind to the top level configuration since it's coming from the container and represents the entire application configuration. It's more likely than not a subsection of that configuration.

```C#
services.AddOptions<MyOptions>()
            .BindConfiguration();
```

It feels like this should look for a section called MyOptions in the configuration.

Thoughts?

Reply:
But isn't this what the `Name` property of the Builder is there for? And as I earlier, we use that multiple other places when binding Configuration to Options.

Reply:
You can have named options for different options types:

```C#
services.AddOptions<MyOptions>("Foo")
            .BindConfiguration();

services.AddOptions<OtherOptions>("Foo")
            .BindConfiguration();
```

Both need to be taken into account. I just think it's extremely rare to ever use the top level configuration. My strawman looks like this:

```C#
services.AddOptions<MyOptions>()
            .BindConfiguration();
```

Looks for the "MyOptions" section:

```C#
services.AddOptions<MyOptions>("Foo")
            .BindConfiguration();
```

Looks for the "Foo:MyOptions" section.

```C#
services.AddOptions<MyOptions>("Foo")
            .BindConfiguration("section");
```

Looks for "section". This overrides any default section selection.

Reply:
Right, my 2c, with the way I've always thought about this binding stuff.  At the end of the day, its about selecting the config section that you want to bind against.  Its not totally unreasonable for there to be situations where the entire configuration file maps to one options class, but that can always be done with the explicit Bind method against the root config section.

That said, are we sure we need to build in a default convention for named options?  Since there's no way to autobind all named options to a specific configuration, I don't think its unreasonable to require them to always specify the path section they want.  So null or empty string as the section path would still enable binding to root.

Reply:
Okay, there are two separate concerns here:

* Binding against the Configuration root
* Omitting a value for the section name and using the `OptionBuilder.Name` for the section name (if none was specified)

So I get the feeling that we want to throw an exception if we do not have a section name (i.e. the section is `null`).

But I feel that using the `Name` of `optionsBuilder` if the `configSectionPath` is `null` is okay? That's just an alternative `configSectionPath`. (However, that could just as well be specified by the caller already if they really wanted to.

``` cs
services.AddOptions<MyOptions>("Foo").BindConfiguration();
```
vs.
``` cs
services.AddOptions<MyOptions>("Foo").BindConfiguration("Foo");
```
for binding the `Foo` config section to `MyOptions`.

And 
``` cs
// Unnamed Builder here
services.AddOptions<MyOptions>().BindConfiguration();
// ^ should throw exception?
```

Reply:
And if you REALLY REALLY wanted to bind against the config root, I suggest that passing an empty string literal would be legal (and not be handled by the code at all)

I.e. this should be allowed (but discouraged and shown in any docs :P ):

``` cs
services.AddOptions<MyOptions>().BindConfiguration(configSectionPath: string.Empty);
```

Reply:
@HaoK well that depends totally on the OptionBuilder name, right? 

Consider:

``` cs 
services.AddOptions<MyOptions>(ConfigurationPath.Combine("AppOptions", "Foo", "Bar")).BindConfiguration();
```

Reply:
Right well that requires them to know the magic of how we bind, I don't think most people are going to name their options like that, I certainly don't haha

Reply:
Yeah, I mostly don't name my options at all, so I am fine with any of this, I was just kinda semi-trying to align with existing implementation in the sister methods declared next to it

Reply:
I have 4 examples there. @fredrikhr can you clarify the behavior you'd expect for those samples? If we end up in a place where there's naming conflicts by in different components by default then it's a bit of failure.

Reply:
@davidfowl

I think an using the type name should not affect the decision for which section is seclected.

Here are my opionions to the four examples:

1.

``` cs 
services.AddOptions<MyOptions>("Foo")
            .BindConfiguration();

services.AddOptions<OtherOptions>("Foo")
            .BindConfiguration();
```
Here we have two builders, both named `Foo`. The developer who wrote this code simply messed up and tries to bind two disjunct options type from the same configuration section. Probably a mistake by the developer, you shouldn't name your options the same, and in this case you probably wanted to specify an actual configuration section path. Unfortunately I don't think there is a reasonable way for us to check this.

Configuration Section chosen: `Foo`

2.

``` cs
services.AddOptions<MyOptions>()
            .BindConfiguration();
```

Unamed builder that does not specify a config section path. Probably not want the developer intended. After thinking about it more, I think we should throw `InvalidOperationException` here, if the developer REALLY wanted to bind against the root, passing empty string as the configuration section path should be legal in the call to `BindConfiguration` (i.e. only `null` is illegal.

Configuration section chosen: `""` (empty string, i.e. the root configuration section)

3. 

``` cs
services.AddOptions<MyOptions>("Foo")
            .BindConfiguration();
```

Named builder, but omitted configuration section path argument.

Configuration section chosen: `Foo`

4.

``` cs
services.AddOptions<MyOptions>("Foo")
            .BindConfiguration("section");
```

Non-null configuration section path argument, binder name is not taken into consideration.

Configuration section chosen: `section`

Reply:
@HaoK @davidfowl I think this discussions shows that there are too many differrent, but all in itself valid/useful opinions on how the configuration section is selected. So my conclusion from this is:

We should enforce a non-null argument for `configurationSectionPath`. And neither the builder name, nor the type name of the builder generic argument are taken into consideration. This would require devs to be explicit about which section in the config they bind against and removes all kind of *magic*  or ambiguous behaviour.

Reply:
The first can't be a bug. Named options are a unique tuple and type and name. I can have 2 different subsystems that look for options worn the same name and different options type and it would work today. It seems unfortunate that this wouldn't work.

Reply:
So I made the change to make `configSectionPath` non-nullable and the only source of truth for the configuration section selection in 3315cf1.

It has the added benefit that the implementation is now much simpler. And easier for callers to understand. I think it is better this way! 😄

Reply:
This is the right call 😁

Reply:
Agree, glad that there's no black box magic with naming conventions by default :)